marketplace/src/lib.rs

```diff
@@ -83,679 +76,514 @@ pub mod pallet {
 			let who = ensure_signed(origin)?;
+			let nft = T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFound)?;
 
-			Marketplaces::<T>::try_mutate(marketplace_id, |x| -> DispatchResult {
-				let market_info = x.as_mut().ok_or(Error::<T>::MarketplaceNotFound)?;
-				ensure!(market_info.owner == who, Error::<T>::NotMarketplaceOwner);
-				market_info.uri = uri.clone();
-				Ok(())
-			})?;
+			// Checks.
+			ensure!(nft.owner == who, Error::<T>::NotTheNFTOwner);
+			ensure!(ListedNfts::<T>::contains_key(nft_id), Error::<T>::NFTNotForSale);
+
+			// Execute.
+			let nft_state = NFTState::new(
+				nft.state.is_capsule,
+				false,
+				nft.state.is_secret,
+				nft.state.is_delegated,
+				nft.state.is_soulbound,
+			);
+			T::NFTExt::set_nft_state(nft_id, nft_state)?;
+			ListedNfts::<T>::remove(nft_id);
+			Self::deposit_event(Event::NFTUnlisted { nft_id });
 
-			let event = Event::MarketplaceUriUpdated { marketplace_id, uri };
-			Self::deposit_event(event);
 			Ok(().into())
 		}
 
-		#[pallet::weight(T::WeightInfo::set_marketplace_logo_uri())]
-		pub fn set_marketplace_logo_uri(
-			origin: OriginFor<T>,
-			marketplace_id: MarketplaceId,
-			logo_uri: URIVec<T>,
-		) -> DispatchResultWithPostInfo {
+		/// Buy a listed nft
+		#[pallet::weight(T::WeightInfo::buy_nft())]
+		#[transactional]
+		pub fn buy_nft(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
+			let nft = T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFound)?;
+			let sale = ListedNfts::<T>::get(nft_id).ok_or(Error::<T>::NFTNotForSale)?;
+			let marketplace = Marketplaces::<T>::get(sale.marketplace_id)
+				.ok_or(Error::<T>::MarketplaceNotFound)?;
+			let mut price = sale.price;
 
-			Marketplaces::<T>::try_mutate(marketplace_id, |x| -> DispatchResult {
-				let market_info = x.as_mut().ok_or(Error::<T>::MarketplaceNotFound)?;
-				ensure!(market_info.owner == who, Error::<T>::NotMarketplaceOwner);
-				market_info.logo_uri = logo_uri.clone();
-				Ok(())
-			})?;
-
-			let event = Event::MarketplaceLogoUriUpdated { marketplace_id, uri: logo_uri };
+			// Checks
+			ensure!(sale.account_id != who, Error::<T>::CannotBuyOwnedNFT);
+			ensure!(T::Currency::free_balance(&who) >= price, Error::<T>::NotEnoughBalanceToBuy);
+
+			// Caller pays for commission fee, the price is updated.
+			let commission_fee = Self::pay_commission_fee(&who, &marketplace, &sale, price)?;
+			price = price.checked_sub(&commission_fee).ok_or(Error::<T>::InternalMathError)?;
+
+			// Caller pays for royalty, the price is updated.
+			let royalty_value = nft.royalty * price;
+			T::Currency::transfer(&who, &nft.creator, royalty_value, KeepAlive)?;
+			price = price.checked_sub(&royalty_value).ok_or(Error::<T>::InternalMathError)?;
+
+			// Caller pays the seller the updated price.
+			T::Currency::transfer(&who, &sale.account_id, price, KeepAlive)?;
+
+			//Execute.
+			let nft_state = NFTState::new(
```

Comment:
I will write it down so that I don't forget. You don't need to create new objects like NFTState or NFTData, you can reuse the same object that you got from line 453